pkg/reconciler/namespacecontroller/namespace_controller.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package namespacecontroller
+
+import (
+	"context"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"kpt.dev/configsync/pkg/status"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// NamespaceController is a controller that watches for Namespace events and
+// sets a flag to indicate whether the Reconciler thread needs to reconcile for
+// this namespace event.
+type NamespaceController struct {
+	client client.Client
+
+	// eventMux is a RMMutex because the cache is read more often than writen.
+	eventMux sync.RWMutex
+
+	// newNSEvent indicates whether the namespace controller fires up a new namespace event.
+	newNSEvent bool
+
+	// selectorMux is the mutex to protect the read/write to nsSelectorMap and selectedNamespaces.
+	selectorMux sync.Mutex
+
+	// nsSelectorMap is a mao of NamespaceSelector names to the LabelSelectors.
+	nsSelectorMap map[string]labels.Selector
+
+	// selectedNamespaces records the dynamic/on-cluster Namespaces selected by NamespaceSelectors
+	selectedNamespaces map[string]struct{}
+}
+
+// New instantiates the namespace controller.
+func New(cl client.Client) *NamespaceController {
+	return &NamespaceController{
+		client:             cl,
+		nsSelectorMap:      make(map[string]labels.Selector),
+		selectedNamespaces: make(map[string]struct{}),
+	}
+}
+
+// setNewNSEvent indicates a new namespace event is detected.
+func (nc *NamespaceController) setNewNSEvent() {
+	nc.eventMux.Lock()
+	defer nc.eventMux.Unlock()
+	nc.newNSEvent = true
+}
+
+// HasNewNSEvent returns whether a new namespace event is detected.
+func (nc *NamespaceController) HasNewNSEvent() bool {
+	nc.eventMux.Lock()
+	defer nc.eventMux.Unlock()
+	return nc.newNSEvent
+}
+
+// SetSelectorCache is invoked by the reconciler to set the cache after parsing the NamespaceSelectors.
+func (nc *NamespaceController) SetSelectorCache(nsSelectorMap map[string]labels.Selector, selectedNamespaces map[string]struct{}) {
```

Comment:
never used?

usually you dont want to set a map like this because maps are mutable. probably want to have individual Clear, Add, Remove methods.

Reply:
It will be used by the follow-up PR when parsing the NamespaceSelector.

Reply:
We don't want to have  ADD/Remove/Clear methods because we update the whole map after parsing all NamespaceSelectors. They should be handled atomically.

Reply:
how about copying the maps then, just to ensure they can't be changed by the caller?

Reply:
done